core/optaplanner-constraint-streams-bavet/src/main/java/org/optaplanner/constraint/streams/bavet/common/AbstractGroupNode.java

```diff
@@ -23,20 +30,30 @@
      * Calls for example {@link AbstractScorer#insert(Tuple)}, and/or ...
      */
     private final TupleLifecycle<OutTuple_> nextNodesTupleLifecycle;
+    /**
+     * Used when {@link #hasMultipleGroups} is true, otherwise {@link #singletonGroup} is used.
+     */
     private final Map<GroupKey_, Group<OutTuple_, GroupKey_, ResultContainer_>> groupMap;
+    /**
+     * Used when {@link #hasMultipleGroups} is false, otherwise {@link #groupMap} is used.
+     */
+    private Group<OutTuple_, GroupKey_, ResultContainer_> singletonGroup;
```

Comment:
This speeds up code paths without grouping. (Such as `groupBy(loadBalance(...))` in `tennis`.)